DuckDuckGo/RemoteMessagingStore.swift

```diff
@@ -0,0 +1,354 @@
+//
+//  RemoteMessagingStore.swift
+//  DuckDuckGo
+//
+//  Copyright © 2022 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import CoreData
+import os.log
+import Core
+import BrowserServicesKit
+
+final class RemoteMessagingStore: RemoteMessagingStoring {
+
+    private enum RemoteMessageStatus: Int16 {
+        case scheduled
+        case dismissed
+        case done
+    }
+
+    private enum Constants {
+        static let privateContextName = "RemoteMessaging"
+        static let remoteMessagingConfigManagedObject = "RemoteMessagingConfigManagedObject"
+        static let remoteMessageManagedObject = "RemoteMessageManagedObject"
+    }
+
+    let notificationCenter: NotificationCenter
+
+    init() {
+        notificationCenter = .default
+    }
+
+    init(context: NSManagedObjectContext, notificationCenter: NotificationCenter = .default) {
+        self.notificationCenter = notificationCenter
+        self.context = context
+    }
+
+    private lazy var context = Database.shared.makeContext(concurrencyType: .privateQueueConcurrencyType, name: Constants.privateContextName)
```

Comment:
Lazy variables are not thread safe, as their initialization can crash when accessed concurrently.

Reply:
That's done